### DISCUSSION TITLE
Compare 2 Filters in a single rule

### DISCUSSION THREAD

**Discussion by noahrotheray** [jertel/elastalert2#1262] [Q&A] [answered]
We are currently trying to migrate all our elastic alerts from AWS lambda to using ElastAlert2.

However we have a few of the alerts that have proven difficult to migrate, specifically ones where we run 2 separate queries for total num_hits, calculate the percentage between the values ( query_1.num_hits / query_2.num_hits *100 ) and send a slack alert if its above X%.

I am currently trying to get this working using `percentage_match`, but I am not sure this is the right approach.
Has anyone managed to get this right or has any ideas?

**Comment by jertel** [comment]
That sounds like the right solution. What problem are you experiencing?

**Comment by noahrotheray** [comment]
Here is the config for my rule

I am trying to send a slack alert when the calculated percentage of: `filter.num_hits / match_bucket_filter.num_hits > 70` in the last 10 minutes.
```
name: "SMS-MT vs DLR"
type: "percentage_match"
index: "logstash-*"
is_enabled: true
realert:
  minutes: 10
terms_size: 50
timeframe:
  minutes: 10
timestamp_field: "@timestamp"
timestamp_type: "iso"
use_strftime_index: False
alert_text_type: alert_text_only
alert_text: "SMS-MT to DLR ratio is less than 70%\n{}"
alert_text_args: 
  - result
min_percentage: 70
doc_type: jasminlogs
match_bucket_filter:
  - bool:
      must: [
            {match: {path: dlrlookupd-messages.log}},
            {match_phrase: {message: "INFO"}},
            {match: {host: smppir2}},
            ]

filter:
  - bool:
      must: [ 
            {match: {path: messages.log}},
            {match_phrase: {message: "SMS-MT"}},
            {match: {host: smppir2}},
            ]
alert:
  - "slack"
```
  
Currently no matter what I set the `min_percentage` it never alerts, I assume this is config related.

I think what I need is a way to test run the alert to see what % is being reported, is there a way to run rules standalone for testing purposes? I am running elastalert2 as a docker container

**Reply by jertel** [reply]
Run the docker with the `--rule <path_to_my_rule_in_the_container> --debug --es_debug` and it will only process your specified rule and use a debug alerter instead of sending the alerts out to slack. You can view the queries and responses to/from Elasticsearch to determine what is actually happening internally.

**Comment by noahrotheray** [comment]
Thanks this has helped alot, however it has shown I might be misunderstanding aggs.

Heres the ES query that runs:
```
POST /logstash-*/_search?pretty&ignore_unavailable=true&size=0
{
  "aggs": {
    "percentage_match_aggs": {
      "filters": {
        "filters": {
          "match_bucket": {
            "bool": {
              "must": [
                {
                  "bool": {
                    "must": [
                      {
                        "match": {
                          "path": "dlrlookupd-messages.log"
                        }
                      },
                      {
                        "match_phrase": {
                          "message": "DLR"
                        }
                      },
                      {
                        "match": {
                          "host": "smppir2"
                        }
                      }
                    ]
                  }
                }
              ]
            }
          }
        },
        "other_bucket": true
      }
    }
  },
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "@timestamp": {
                  "gt": "2023-09-19T13:00:18.423804Z",
                  "lte": "2023-09-19T13:10:18.423804Z"
                }
              }
            },
            {
              "bool": {
                "must": [
                  {
                    "match": {
                      "path": "messages.log"
                    }
                  },
                  {
                    "match_phrase": {
                      "message": "SMS-MT"
                    }
                  },
                  {
                    "match": {
                      "host": "smppir2"
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    }
  }
}
```

And the result:

```
{
  "took": 51,
  "timed_out": false,
  "_shards": {
    "total": 81,
    "successful": 81,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 10000,
      "relation": "gte"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "percentage_match_aggs": {
      "buckets": {
        "match_bucket": {
          "doc_count": 0
        },
        "_other_": {
          "doc_count": 18916
        }
      }
    }
  }
}
```
the doc_count for the match_bucket is 0, however if i run the match_bucket query on its own, there are definitely hits. 

Is this because the match_bucket is looking for matches within the results/docs of the primary filter? Can I actually use 2 completely different queries as these use 2 separate log files (messages.log and dlrlookupd-messages.log), there will never be any actual matches between the docs? I just want to compare the total hits between them to calculate a percentage.

**Comment by jertel** [comment]
Yes, the numerator is a subset of the denominator. From the `[docs](https://elastalert2.readthedocs.io/en/latest/ruletypes.html#percentage-match)`:

![image](https://github.com/jertel/elastalert2/assets/16639611/c8cf34c1-17c3-43af-900e-e407f8c4397a)

I'm not aware of a way to compare two distinct results sets in ElastAlert 2. Perhaps you could have an Elastic scheduled job perform the query and comparison and log the results into a new index which ElastAlert 2 is monitoring.

Or change the rule to compare count of INFO logs to count of total logs.

**Comment by noahrotheray** [comment]
Good idea, I managed to create a new index using Transform. This now has 2 fields: `dlrs` and `mts` which seems to have made thingseasier. Example doc:

```
      {
        "_index": "sms_dlrs_vs_mts",
        "_id": "AGn2MYjwOLjneyOYrHiYTGIAAAAAAAAA",
        "_score": 0,
        "_source": {
          "@timestamp": "2023-07-20T12:08:00.000Z",
          "dlrs": 287,
          "mts": 391
        }

```
I'm now trying to use the  `metric_aggregation` with `metric_agg_script`

```
name: "SMS-MT vs DLR"
type: "metric_aggregation"
index: "sms_dlrs_vs_mts"
is_enabled: true
realert:
  minutes: 10
terms_size: 50
timeframe:
  minutes: 10
timestamp_field: "@timestamp"
timestamp_type: "iso"
use_strftime_index: False
alert_text_type: alert_text_only
alert_text: "SMS-MT to DLR ratio is less than 70%\n{}"
alert_text_args:
  - percentage
min_threshold: 0.99
metric_agg_type: max
metric_agg_key: delivery_percentage
metric_agg_script:
    script: doc['dlrs'].value / doc['mts'].value * 100
```

ES query:

```
{
  "aggs": {
    "metric_delivery_percentage_max": {
      "max": {
        "script": "doc['mts'].value / doc['dlrs'].value * 100"
      }
    }
  },
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "@timestamp": {
                  "gt": "2023-07-20T12:07:14.256621Z",
                  "lte": "2023-07-20T12:22:14.256621Z"
                }
              }
            }
          ]
        }
      }
    }
  }
}
```

Not sure if this is working the way I think it is though, if I run the query myself, all I get is:

```
      {
        "_index": "sms_dlrs_vs_mts",
        "_id": "ANM2eY9T94sQE7lwJhceKWIAAAAAAAAA",
        "_score": 0,
        "_source": {
          "@timestamp": "2023-07-20T12:17:00.000Z",
          "dlrs": 244,
          "mts": 236
        }
      }
    ]
  },
  "aggregations": {
    "metric_delivery_percentage_max": {
      "value": 100
    }
  }
}
```

Would this be the right approach now that I have the 2 fields i want to calculate in the same doc and index? Also, I assumed filter isnt neccesary here as I just want the result of the `metric_agg_script`

**Reply by jertel** [reply]
Yes I think this is a good approach. I wonder if you have a type/rounding issue. 244/236 = 1.033 * 100 = 103.3. But if the script engine is casting those values as ints then perhaps 244 / 236 = 1 * 100 = 100.

**Comment by noahrotheray** [comment, marked as answer]
Thanks  @jertel, but i think I found an easier to do this, again using transform :)

I created a new transform that calculates the % using an aggregation in the transform JSON, so its now just throwing the field into the index, which I can query with elastalert2, below is the transform JSON:

```
"pivot": {
    "group_by": {
      "@timestamp": {
        "date_histogram": {
          "field": "@timestamp",
          "fixed_interval": "10m"
        }
      }
    },
    "aggregations": {
      "dlrs": {
        "sum": {
          "field": "dlrs"
        }
      },
      "mts": {
        "sum": {
          "field": "mts"
        }
      },
      "delivery_percentage": {
        "bucket_script": {
          "buckets_path": {
            "dlrs": "dlrs",
            "mts": "mts"
          },
          "script": "params.dlrs / params.mts * 100"
        }
      }
    }
  },
```
So far this seems to doing exactly what I need, appreciate the help!
